File: tyama_codeiq684_fast.rb
#!/usr/bin/ruby
#gemstring fast solution
#based on https://gist.github.com/tatt61880/8958003

=begin
class Array

	# yields all Arrays with enumerating "all elements from each elements".
	#
	#   [[1,2,3],[4,5,6]].choices.to_a
	#   #=> [[1, 4], [1, 5], [1, 6], [2, 4], [2, 5], [2, 6], [3, 4], [3, 5], [3, 6]]

	def choices(a=[],d=0,&block)
		return to_enum(:choices,[],0) if !block_given?
		if d==self.size
			yield a
		else
			self[d].size.times{|i|
				choices(a+[self[d][i]],d+1,&block)
			}
		end
	end

end
=end

class Integer
	def fact() (1..self).reduce(1,:*) end
end

counts=Hash.new(0)
'abbbbcddddeefggg'.chars{|c|counts[c.ord-'a'.ord]+=1}
r=0
desired='eagcdfbe'.chars.map{|c|c.ord-'a'.ord}
desired.each{|n|
	r+=1
	n.times{|i|
		counts[i]-=1
		val=counts.values.map{|e|[*0..e]}
		#val.choices{|e|
		val[0].product(*val[1..-1]).each{|e|
			r+=e.reduce(:+).fact / e.map(&:fact).reduce(:*)
		}
		counts[i]+=1
	}
	counts[n]-=1
}
p r

File: codeiq/tyama_codeiq684_fast.cpp
//gemstring fast solution
//based on https://gist.github.com/tatt61880/8958003

#include <vector>
#include <cstdio>
#include <cstring>
using namespace std;
#define list     "abbbbcddddeefggg"
#define desired_ "eagcdfbe"

vector<long long>fact;
long long dfs(vector<int>&v,vector<int> &a,int d){
	long long r=0;
	if(d==v.size()){
		for(int i=0;i<a.size();i++)r+=a[i];
		r=fact[r];
		for(int i=0;i<a.size();i++)r/=fact[a[i]];
	}else{
		for(int i=0;i<=v[d];i++){
			a.push_back(i);
			r+=dfs(v,a,d+1);
			a.pop_back();
		}
	}
	return r;
}

int main(){
	vector<int> a;
	vector<int> counts(7); //should not be hard-coded...
	for(int i=0;i<strlen(list);i++)counts[list[i]-'a']++;
	fact.push_back(1);
	for(int i=1;i<strlen(list);i++)fact.push_back(fact[i-1]*i);
	vector<int> desired(strlen(desired_));
	for(int i=0;i<strlen(desired_);i++)desired[i]=desired_[i]-'a';

	long long r=0;
	for(int n=0;n<strlen(desired_);n++){
		r++;
		for(int i=0;i<desired[n];i++){
			counts[i]--;
			r+=dfs(counts,a,0);
			counts[i]++;
		}
		counts[desired[n]]--;
	}
	printf("%lld\n",r);
}

/*
naive:
Ruby unknown
C#   unknown
C    21s
fast:
Ruby 0.22s
C++  0.01s
*/